libkbfs/backpressure_disk_limiter.go

```diff
@@ -69,71 +83,98 @@ func defaultDoDelay(ctx context.Context, delay time.Duration) error {
 // newBackpressureDiskLimiter constructs a new backpressureDiskLimiter
 // with the given parameters.
 func newBackpressureDiskLimiter(
-	backpressureMinThreshold, backpressureMaxThreshold, byteLimit int64,
-	maxDelay time.Duration) backpressureDiskLimiter {
-	return newBackpressureDiskLimiterWithDelayFunction(
+	backpressureMinThreshold, backpressureMaxThreshold float64,
+	byteLimit int64, maxDelay time.Duration,
+	journalPath string) *backpressureDiskLimiter {
+	return newBackpressureDiskLimiterWithFunctions(
 		backpressureMinThreshold, backpressureMaxThreshold,
-		byteLimit, maxDelay, defaultDoDelay)
+		byteLimit, maxDelay, defaultDoDelay, func() (int64, error) {
+			availBytes, err := getDiskLimits(journalPath)
+			if err != nil {
+				return 0, err
+			}
+
+			if availBytes > uint64(math.MaxInt64) {
+				return math.MaxInt64, nil
+			}
+			return int64(availBytes), nil
+		})
 }
 
-func (s backpressureDiskLimiter) onJournalEnable(journalBytes int64) int64 {
-	if journalBytes == 0 {
-		return s.s.Count()
-	}
-	return s.s.ForceAcquire(journalBytes)
+func (s *backpressureDiskLimiter) onJournalEnable(journalBytes int64) int64 {
+	s.journalBytesLock.Lock()
+	defer s.journalBytesLock.Unlock()
+	s.journalBytes += journalBytes
+	return 0
 }
 
-func (s backpressureDiskLimiter) onJournalDisable(journalBytes int64) {
-	if journalBytes > 0 {
-		s.s.Release(journalBytes)
-	}
+func (s *backpressureDiskLimiter) onJournalDisable(journalBytes int64) {
+	s.journalBytesLock.Lock()
+	defer s.journalBytesLock.Unlock()
+	s.journalBytes -= journalBytes
 }
 
-func (s backpressureDiskLimiter) getDelay() time.Duration {
-	availBytes := s.s.Count()
-	usedBytes := s.byteLimit - availBytes
-	if usedBytes <= s.backpressureMinThreshold {
-		return 0
-	}
-
-	if usedBytes >= s.backpressureMaxThreshold {
-		return s.maxDelay
-	}
+func boundedLerp(x, min, max float64) float64 {
+	return math.Min(1.0, math.Max(0.0, (x-min)/(max-min)))
+}
 
-	scale := float64(usedBytes-s.backpressureMinThreshold) /
-		float64(s.backpressureMaxThreshold-s.backpressureMinThreshold)
-	delayNs := int64(float64(s.maxDelay.Nanoseconds()) * scale)
-	return time.Duration(delayNs) * time.Nanosecond
+func (s *backpressureDiskLimiter) getDelay(availBytes int64) time.Duration {
+	journalBytes := func() int64 {
+		s.journalBytesLock.RLock()
+		defer s.journalBytesLock.RUnlock()
+		return s.journalBytes
+	}()
+
+	r := math.Max(
+		float64(journalBytes)/float64(s.maxJournalBytes),
+		float64(journalBytes)/(float64(journalBytes)+float64(availBytes)))
+	m := s.backpressureMinThreshold
+	M := s.backpressureMaxThreshold
+	scale := math.Min(1.0, math.Max(0.0, (r-m)/(M-m)))
+	return time.Duration(scale * float64(s.maxDelay))
```

Comment:
If we wait a full `maxDelay` on every block of a multi-block put, we'll run into OS timeouts pretty quickly.  However, since blocks are mostly put in parallel, outside of the journal lock, I guess in practice we're actually waiting only one `maxDelay` as long as there aren't more than 100 blocks put at once.

Still, maybe it's safer to pass in `ctx` here and scale  `min(s.maxDelay, ctx.Deadline().Sub(time.Now())-1*time.Second)` or something similar.